src/bvals/comms/coalesced_buffers.cpp

```diff
@@ -0,0 +1,482 @@
+  var_buf_t *pbuf = &(pmesh->boundary_comm_map.at(key));
+  coalesced_info_buf[bnd_id.var_id()].push_back(std::make_pair(bnd_id, pbuf));
+  current_size += bnd_id.size(); // This will be the maximum size of communication since
+                                 // it includes all variables
+  all_vars.insert(bnd_id.var_id());
+}
+
+void CoalescedBuffer::AddVarBoundary(MeshBlock *pmb, const NeighborBlock &nb,
+                                     const std::shared_ptr<Variable<Real>> &var) {
+  // Store both the variable-boundary buffer information and a pointer to the v-b buffer
+  // itself associated with var ids
+  BndId bnd_id = BndId::GetSend(pmb, nb, var, b_type, partition, -1);
+  AddVarBoundary(bnd_id);
+}
+
+//----------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------
+//----------------------------------------------------------------------------------------
+CoalescedBuffersRank::CoalescedBuffersRank(int o_rank, BoundaryType b_type, bool send,
+                                           mpi_comm_t comm, Mesh *pmesh)
+    : other_rank(o_rank), b_type(b_type), sender(send), buffers_built(false), comm_(comm),
+      pmesh(pmesh) {
+
+  int tag = 1234 + static_cast<int>(GetAssociatedSender(b_type));
+  if (sender) {
+    message = com_buf_t(tag, Globals::my_rank, other_rank, comm_,
+                        [](int size) { return std::vector<int>(size); });
+  } else {
+    message = com_buf_t(
+        tag, other_rank, Globals::my_rank, comm_,
+        [](int size) { return std::vector<int>(size); }, true);
+  }
+  PARTHENON_REQUIRE(other_rank != Globals::my_rank, "Should only build for other ranks.");
+}
+
+//----------------------------------------------------------------------------------------
+void CoalescedBuffersRank::AddSendBuffer(int partition, MeshBlock *pmb,
+                                         const NeighborBlock &nb,
+                                         const std::shared_ptr<Variable<Real>> &var) {
+  if (coalesced_bufs.count(partition) == 0)
+    coalesced_bufs.emplace(
+        std::make_pair(partition, CoalescedBuffer(true, partition, other_rank, b_type,
+                                                  comm_, pmb->pmy_mesh)));
+
+  auto &coal_buf = coalesced_bufs.at(partition);
+  coal_buf.AddVarBoundary(pmb, nb, var);
+}
+
+//----------------------------------------------------------------------------------------
+bool CoalescedBuffersRank::TryReceiveBufInfo() {
+  PARTHENON_REQUIRE(!sender, "Trying to receive on a combined sender.");
+  if (buffers_built) return buffers_built;
+
+  bool received = message.TryReceive();
+  if (received) {
+    auto &mess_buf = message.buffer();
+    int npartitions = mess_buf[0];
+    // Unpack into per combined buffer information
+    int idx{nglobal};
+
+    for (int p = 0; p < npartitions; ++p) {
```

Comment:
Not related to this piece of code but a more general question/comment. Do we still support arbitary partitions during runtime (so except for the "all blocks" and the "default split based on `pack_size`")?

Reply:
This PR only allows for coalesced communication on the default partition (i.e. the one determined by `pack_size`). That is because there is a single coalesced message (plus a message for sparse info) for each `MeshData`. The contents of these messages need to be determined and sent to neighbor ranks during remeshing and we only do this for the default partitions. I am guessing we could add communication information for other partitions if we really wanted to at the same time, but there are probably some subtle gotchas in there. 

Of course, I don't think this prevents you from setting up a different partitioning that you don't communicate on. Also, I think the issues with this PR we are seeing in Riot are related to an issue with the "all blocks" partition (i.e. it doesn't get a partition id assigned). It is probably necessary to generalize partition ids to include the all blocks partition (as well as other possible partitions maybe).